explorer/common/arena.h

```diff
@@ -61,35 +119,162 @@ class Arena {
 
   // Templated destruction of a pointer.
   template <typename T>
-  class ArenaEntryTyped : public ArenaEntry {
-   public:
-    struct WriteAddressHelper {};
+  class ArenaEntryTyped;
 
-    template <typename... Args>
-    explicit ArenaEntryTyped(Args&&... args)
-        : instance_(std::forward<Args>(args)...) {}
+  // Manages allocations in an arena for destruction at shutdown.
+  std::vector<std::unique_ptr<ArenaEntry>> arena_;
+  int64_t allocated_ = 0;
 
-    template <typename... Args>
-    explicit ArenaEntryTyped(WriteAddressHelper, Args&&... args)
-        : ArenaEntryTyped(std::forward<Args>(args)...) {}
+  // Hash functor implemented in terms of hash_value (see llvm/ADT/Hashing.h)
+  struct LlvmHasher {
+    template <typename T>
+    auto operator()(const T& t) const -> size_t {
+      using llvm::hash_value;
+      return hash_value(t);
+    }
+  };
 
-    template <typename U, typename... Args>
-    explicit ArenaEntryTyped(WriteAddressTo<U> write_address, Args&&... args)
-        : ArenaEntryTyped(
-              (*write_address.target = &instance_, WriteAddressHelper{}),
-              std::forward<Args>(args)...) {}
+  // Factory metafunction for globally unique type IDs.
+  // &TypeId<T>::id == &TypeId<U>::id if and only if std::is_same_v<T,U>.
+  //
+  // Inspired by llvm::Any::TypeId.
+  template <typename T>
+  struct TypeId {
+    static char id;
+  };
+
+  // A canonicalization table maps a tuple of constructor argument values to
+  // a non-null pointer to a T object constructed with those arguments.
+  template <typename T, typename... Args>
+  using CanonicalizationTable =
+      std::unordered_map<std::tuple<ArgKeyType<Args>...>, Nonnull<const T*>,
+                         LlvmHasher>;
```

Comment:
Did you consider using llvm::FoldingSet for this? We're using it in the toolchain now for type deduplication (in SemanticsContext), and I'm wondering if it'd be a reasonable choice here too, although the approach would be a bit different for how hashing is done versus if you were to build a FoldingSetNodeID.